File: backend/app/schemas/prescription.py
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field


class PrescriptionBase(BaseModel):
    customer_id: int
    prescription_date: date
    right_sph: Decimal | None = None
    right_cyl: Decimal | None = None
    right_axis: int | None = Field(default=None, ge=0, le=180)
    right_vn: str | None = Field(default=None, max_length=20)
    left_sph: Decimal | None = None
    left_cyl: Decimal | None = None
    left_axis: int | None = Field(default=None, ge=0, le=180)
    left_vn: str | None = Field(default=None, max_length=20)
    fh: str | None = Field(default=None, max_length=32)
    add_power: Decimal | None = None
    pd: Decimal | None = None
    notes: str | None = None


class PrescriptionCreate(PrescriptionBase):
    pass


class PrescriptionUpdate(BaseModel):
    prescription_date: date | None = None
    right_sph: Decimal | None = None
    right_cyl: Decimal | None = None
    right_axis: int | None = Field(default=None, ge=0, le=180)
    right_vn: str | None = Field(default=None, max_length=20)
    left_sph: Decimal | None = None
    left_cyl: Decimal | None = None
    left_axis: int | None = Field(default=None, ge=0, le=180)
    left_vn: str | None = Field(default=None, max_length=20)
    fh: str | None = Field(default=None, max_length=32)
    add_power: Decimal | None = None
    pd: Decimal | None = None
    notes: str | None = None


class PrescriptionRead(PrescriptionBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    created_at: datetime
    updated_at: datetime
    created_by: int | None
    updated_by: int | None
    is_deleted: bool
    customer_name: str | None = None
    customer_business_id: str | None = None
    customer_contact_no: str | None = None


class PrescriptionListResponse(BaseModel):
    items: list[PrescriptionRead]
    total: int
    page: int
    page_size: int


class PrescriptionPdfResponse(BaseModel):
    prescription_id: int
    pdf_url: str


class PrescriptionSendVendorRequest(BaseModel):
    vendor_id: int = Field(ge=1)
    caption: str | None = Field(default=None, max_length=500)


class PrescriptionSendVendorResponse(BaseModel):
    message: str
    whatsapp_log_id: int | None = None
    provider_message_id: str | None = None

File: frontend/src/types/prescription.ts
export type Prescription = {
  id: number;
  customer_id: number;
  customer_name: string | null;
  customer_business_id: string | null;
  customer_contact_no: string | null;
  prescription_date: string;
  right_sph: number | null;
  right_cyl: number | null;
  right_axis: number | null;
  right_vn: string | null;
  left_sph: number | null;
  left_cyl: number | null;
  left_axis: number | null;
  left_vn: string | null;
  fh: string | null;
  add_power: number | null;
  pd: number | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
  created_by: number | null;
  updated_by: number | null;
  is_deleted: boolean;
};

export type PrescriptionPayload = {
  customer_id: number;
  prescription_date: string;
  right_sph?: number | null;
  right_cyl?: number | null;
  right_axis?: number | null;
  right_vn?: string | null;
  left_sph?: number | null;
  left_cyl?: number | null;
  left_axis?: number | null;
  left_vn?: string | null;
  fh?: string | null;
  add_power?: number | null;
  pd?: number | null;
  notes?: string | null;
};

export type PrescriptionPdfResponse = {
  prescription_id: number;
  pdf_url: string;
};

export type PrescriptionSendVendorPayload = {
  vendor_id: number;
  caption?: string | null;
};

export type PrescriptionSendVendorResponse = {
  message: string;
  whatsapp_log_id: number | null;
  provider_message_id: string | null;
};
